IronFrame.Shared/Utilities/JobObject.cs

```diff
@@ -332,13 +361,82 @@ public virtual void TerminateProcessesAndWait(int milliseconds = System.Threadin
         // P/Invoke Helpers
         //
 
-        static Exception Win32LastError(string message, params object[] args)
+        private static Exception Win32LastError(string message, params object[] args)
         {
             var error = Marshal.GetLastWin32Error();
             return new Win32Exception(
-                error, 
-                String.Format(message, args) + 
-                    String.Format(" (Win32 Error Code {0})", error));
+                error,
+                String.Format(message, args) +
+                String.Format(" (Win32 Error Code {0})", error));
+        }
+
+
+        public void SetJobCpuLimit(int weight)
+        {
+            if (weight < 1 || weight > 9)
+            {
+                throw new ArgumentOutOfRangeException("weight", weight, "CPU Limit must be between 1 and 9");
+            }
+            int length = Marshal.SizeOf(typeof(NativeMethods.JobObjectCpuRateControlInformation));
+            IntPtr cpuRatePtr = IntPtr.Zero;
+
+            try
+            {
+                var cpuRate = new NativeMethods.JobObjectCpuRateControlInformation
+                {
+                    ControlFlags = (UInt32)(NativeMethods.JobObjectCpuRateControl.Enable | NativeMethods.JobObjectCpuRateControl.WeightBased),
+                    Weight = (UInt32)weight,
+                };
+
+                cpuRatePtr = Marshal.AllocHGlobal(length);
```

Comment:
We haven't been very good at updating existing code to use consistent patterns, but there is a helper that you should use here:

```
using (var allocation = SafeAllocation.Create(cpuRate))
{
    if (!NativeMethods...)
    {
    }
}
```

The `SafeAllocation` helper will allocate the structure and marshal the managed object into the memory buffer, and then automatically free it at the end of the using block.

There are a couple of other examples of using `SafeAllocation`.